partiql-ast/src/main/java/org/partiql/ast/v1/DataType.java

```diff
@@ -391,6 +396,109 @@ public int code() {
         return code;
     }
 
+    public static DataType valueOf(String value) {
+        switch (value) {
+            case "NULL": return NULL();
+            case "MISSING": return MISSING();
+            case "BOOL": return BOOL();
+            case "BOOLEAN": return BOOLEAN();
+            case "TINYINT": return TINYINT();
+            case "SMALLINT": return SMALLINT();
+            case "INTEGER2": return INTEGER2();
+            case "INT2": return INT2();
+            case "INTEGER": return INTEGER();
+            case "INT": return INT();
+            case "INTEGER4": return INTEGER4();
+            case "INT4": return INT4();
+            case "INTEGER8": return INTEGER8();
+            case "INT8": return INT8();
+            case "BIGINT": return BIGINT();
+            case "REAL": return REAL();
+            case "DOUBLE_PRECISION": return DOUBLE_PRECISION();
+            case "FLOAT": return FLOAT();
+            case "DECIMAL": return DECIMAL();
+            case "DEC": return DEC();
+            case "NUMERIC": return NUMERIC();
+            case "BIT": return BIT();
+            case "BIT_VARYING": return BIT_VARYING();
+            case "CHAR": return CHAR();
+            case "CHARACTER": return CHARACTER();
+            case "VARCHAR": return VARCHAR();
+            case "CHARACTER_LARGE_OBJECT": return CHARACTER_LARGE_OBJECT();
+            case "CHAR_LARGE_OBJECT": return CHAR_LARGE_OBJECT();
+            case "CHAR_VARYING": return CHAR_VARYING();
+            case "STRING": return STRING();
+            case "SYMBOL": return SYMBOL();
+            case "BLOB": return BLOB();
+            case "BINARY_LARGE_OBJECT": return BINARY_LARGE_OBJECT();
+            case "CLOB": return CLOB();
+            case "DATE": return DATE();
+            case "STRUCT": return STRUCT();
+            case "TUPLE": return TUPLE();
+            case "LIST": return LIST();
+            case "SEXP": return SEXP();
+            case "BAG": return BAG();
+            case "TIME": return TIME();
+            case "TIME_WITH_TIME_ZONE": return TIME_WITH_TIME_ZONE();
+            case "TIMESTAMP": return TIMESTAMP();
+            case "TIMESTAMP_WITH_TIME_ZONE": return TIMESTAMP_WITH_TIME_ZONE();
+            case "INTERVAL": return INTERVAL();
+            case "USER_DEFINED": return USER_DEFINED();
+            default: return UNKNOWN();
+        }
+    }
+
+    public static DataType[] values() {
```

Comment:
I suggest these are all singletons, we want `enum.FOO() === enum.FOO()` – aka they return the same reference.

Reply:
From offline discussion, we want separate instances since some enum nodes will extend `AstNode` in subsequent PRs. Those nodes like `DataType` may need source locations which require a tag in the parser's source location map.